Holograms

Pspritechologist opened the pull request:
# Description
Adds Holograms, an early game alternative to cloning. Coming back as a Hologram is one part trans humanist nightmare, one part mockery of the very fundamental aspects of life, and 100% not being alive!


Will describe more once features are done.
Soon :tm: 

---

# Tasks

- [ ] Task
- [x] Completed Task

---

<details><summary><h1>Media</h1></summary>
<p>

<!-- ![Example Media Embed](https://example.com) -->

</p>
</details>

---

# Changelog

:cl:
- add: Added fun!
- tweak: Tweaked fun!
- fix: Fixed fun!
- remove: Removed fun!

Pspritechologist committed:
Added a new mob typed named Hologram, set up a clone of the Drone whitelist system for "hardlight

Pspritechologist committed:
adds early scutters, and the basis of a charging system.

Pspritechologist committed:
Working on it

Pspritechologist committed:
Lots more work towards holograms, probably too long without committing :P

Pspritechologist committed:
Committing more

Pspritechologist committed:
More work on Holograms. Moved most things over to events, bug where hologram is deleted upon spawning and crashes as no hologram component is found?

Pspritechologist committed:
Fixed holocorgi ID

Pspritechologist committed:
Fuckin working again. Stuff is networked now, though not to any real benefit. There's a grace period upon leaving a projector of one 24th of a second before you get returned.

Pspritechologist committed:
Prepared for projected, lightbee, and hardlight holos by separating the protos, make a temp new tag Light, deleted unused files, renamed HoloData to HoloMind, and made HoloData struct to hold info on the Holo like type and whether it's hardlight. Needs implementing in YAML.

Pspritechologist committed:
Removed some unneeded things, cleaned some files, made it Holograms can interact with light things, and non-Holograms can't.

Pspritechologist committed:
Added damage sets for Holograms, kinda fixed rotation on spawn, begun work on hologram data in the component (no structs ;-;), and refined the variations of Holograms.

Pspritechologist committed:
Removed unneeded files

Pspritechologist committed:
Removed unused component

Pspritechologist committed:
Death's tweaks, filtered :)

Pspritechologist committed:
Made Enum worked, cleaned up a bit, and fixed all namespaces

Pspritechologist committed:
Substantial cleanup to the code post rebase

Man I was really smashing rocks together before :P
Also I removed the stinky Nyano Holo stuff

Pspritechologist committed:
Fixes double-playing audio

Pspritechologist committed:
Completely overhauled large portions.

The HologramComponent has now been broken up into three Components, soon to be more. All the functions of the HologramSystem are chunked up and completely reconsidered.
Prediction is cool.

Pspritechologist committed:
Bulk of work done towards the HologramProjectedComponent's functionality.

Need to decide how the projector selection is done, and how to filter collisions.

Pspritechologist committed:
Ton of work refactoring Holograms, generally cleaning it up, and adding new functionality.

The AIEye can now function as a camera-bound Hologram (commented out), work still needs to be done on making Hologram disks and servers cool to use, and I need to decide how to determine what a Hologram collides with.

Pspritechologist commented on Resources/Prototypes/SimpleStation14/Entities/Mobs/Player/stationai.yml:
```diff
@@ -325,7 +332,7 @@
     # Can see through walls for now
     - type: Eye
-      drawFov: false
+      # drawFov: true
```
Whoops, commenting this out won't have the desired effect :P

Pspritechologist commented on Resources/Prototypes/SimpleStation14/Entities/Mobs/Player/stationai.yml:
```diff
@@ -227,6 +227,13 @@
+    # - type: HologramServer
+    # - type: HologramProjector
+    #   effectOffsets:
+    #     North: "0, 0.40"
+    #     East: "-0.15, 0"
+    #     South: "0, -0.20"
+    #     West: "-0.15, 0"
```
I think these are upside down?

Pspritechologist commented on Resources/Prototypes/SimpleStation14/Entities/Mobs/Player/silicon.yml:
Scutters haven't been updated along with other stuff, and they'll probably want most of a redo in general.

Pspritechologist commented on Resources/Prototypes/SimpleStation14/Entities/Mobs/NPCs/pets.yml:
```diff
@@ -45,3 +45,77 @@
+- type: entity
+  name: holo corgi
+  description: "A hologramatic projection of a corgi, computed by the SAI and rendered by the station's cameras."
+  id: MobCorgiHolo
```
Ditto for HoloCorgi as Scutter

Pspritechologist commented on Resources/Prototypes/SimpleStation14/Body/Prototypes/hologram.yml:
Should probably just have human bodies? I think they might...

Pspritechologist commented on Resources/Prototypes/Entities/Structures/Machines/Computers/base_structurecomputers.yml:
```diff
@@ -57,3 +57,6 @@
         ents: []
+  - type: Tag
+    tags:
+    - Hardlight
```
So that Holograms can interact with computers... Really not ideal, though.

DEATHB4DEFEAT replied:
They can't do that!
*The sAI can't*

Pspritechologist commented on Resources/Prototypes/Entities/Structures/Doors/Airlocks/base_structureairlocks.yml:
```diff
@@ -131,5 +131,8 @@
   - type: DynamicPrice
     price: 150
   - type: AccessReader
+  - type: Tag
```
Probably not needed, maybe good to have anyway? I could change collision to be an EntityWhitelist instead of tags and then just use the Component

Pspritechologist commented on Resources/Prototypes/Entities/Objects/Tools/glowstick.yml:
```diff
@@ -61,6 +61,9 @@
           property: Radius
+    - type: Tag
+      tags:
+        - Hardlight # it is a hard light
```
funny

Pspritechologist commented on Resources/Audio/SimpleStation14/Effects/Hologram/holo_on.ogg:
People don't like these I guess? I dunno, I'm fine with them, if anyone wants to get anything better feel free

Pspritechologist committed:
Merge branch 'master' into holograms

DEATHB4DEFEAT commented on Content.Client/SimpleStation14/Holograms/HologramSystem.cs:
```diff
@@ -0,0 +1,129 @@
+{
+    [Dependency] private readonly IPlayerManager _player = default!;
+    [Dependency] private readonly TransformSystem _transform = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<HologramProjectedComponent, ComponentShutdown>(OnProjectedShutdown);
+    }
+
```
```suggestion


    public override void Initialize()
    {
        base.Initialize();

        SubscribeLocalEvent<HologramProjectedComponent, ComponentShutdown>(OnProjectedShutdown);
    }


```

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Holograms/Systems/HologramProjectorSystem.cs:
```diff
@@ -0,0 +1,34 @@
+        SubscribeLocalEvent<HologramProjectorComponent, SurveillanceCameraChangeStateEvent>((ent, comp, args) => CheckState(ent, comp));
+        SubscribeLocalEvent<HologramProjectorComponent, MapInitEvent>((ent, comp, args) => CheckState(ent, comp));
+    }
+
```
```suggestion


```

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Holograms/Systems/HologramServerSystem.cs:
```diff
@@ -0,0 +1,139 @@
+        SubscribeLocalEvent<HologramDiskComponent, AfterInteractEvent>(OnAfterInteract);
+        SubscribeLocalEvent<HologramServerComponent, PowerChangedEvent>(OnPowerChanged);
+    }
+
```
```suggestion


    public override void Initialize()
    {
        base.Initialize();
        SubscribeLocalEvent<HologramServerComponent, EntInsertedIntoContainerMessage>(OnEntInserted);
        SubscribeLocalEvent<HologramServerComponent, EntRemovedFromContainerMessage>(OnEntRemoved);
        SubscribeLocalEvent<HologramDiskComponent, AfterInteractEvent>(OnAfterInteract);
        SubscribeLocalEvent<HologramServerComponent, PowerChangedEvent>(OnPowerChanged);
    }


```

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Holograms/Systems/HologramServerSystem.cs:
```diff
@@ -0,0 +1,139 @@
+
+        if (Exists(component.LinkedHologram))
+            if (!_hologram.TryKillHologram(component.LinkedHologram.Value))
+                return; // This is a werid situation to encounter, so we'll just stop doin stuff.
```
```suggestion
                return; // This is a weird situation to encounter, so we'll just stop doin stuff.
```

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Holograms/Systems/HologramSystem.cs:
```diff
@@ -0,0 +1,190 @@
+    [Dependency] private readonly StationSystem _station = default!;
+
+    public readonly Dictionary<Mind.Mind, EntityUid> HologramsWaitingForMind = new();
+
```
```suggestion

```

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Holograms/Systems/HologramSystem.cs:
```diff
@@ -0,0 +1,190 @@
+        if (TryComp<MindContainerComponent>(hologram, out var mindComp) && mindComp.Mind != null)
+            _gameTicker.OnGhostAttempt(mindComp.Mind, false);
+
+        _audio.Play(filename: "/Audio/SimpleStation14/Effects/Hologram/holo_off.ogg", playerFilter: Filter.Pvs(hologram), coordinates: holoPos, false);
```
Shouldn't be hard coded

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Holograms/Systems/HologramSystem.cs:
```diff
@@ -0,0 +1,190 @@
+        _adminLogger.Add(LogType.Mind, LogImpact.Medium, $"{ToPrettyString(hologram):mob} was killed!");
+    }
+
+    public bool TryGenerateHologram(Mind.Mind mind, EntityCoordinates coords, [NotNullWhen(true)] out EntityUid? holo)
```
dox

DEATHB4DEFEAT commented on Content.Server/SurveillanceCamera/Systems/SurveillanceCameraSystem.cs:
```diff
@@ -58,7 +58,7 @@ public override void Initialize()
         SubscribeLocalEvent<SurveillanceCameraComponent, SurveillanceCameraSetupSetNetwork>(OnSetNetwork);
         SubscribeLocalEvent<SurveillanceCameraComponent, GetVerbsEvent<AlternativeVerb>>(AddVerbs);
-        
+
```
```suggestion
        
```

DEATHB4DEFEAT commented on Content.Shared/SimpleStation14/Holograms/Components/HologramProjectedComponent.cs:
```diff
@@ -0,0 +1,92 @@
+    /// <summary>
+    ///     A whitelist to check for on projectors, to determine if they're valid.
+    /// </summary>
+    [DataField("validProjectorWhitelist"), ViewVariables(VVAccess.ReadWrite)]
```
```suggestion
    [DataField("validProjectorWhitelist"), ViewVariables(VVAccess.ReadWrite)]
    [AutoNetworkedField]
```

DEATHB4DEFEAT commented on Content.Shared/SimpleStation14/Holograms/Components/HologramProjectedComponent.cs:
```diff
@@ -0,0 +1,92 @@
+    /// <summary>
+    ///     A timer for a grace period before the Holo is returned, to allow for moving through doors.
+    /// </summary>
+    [DataField("gracePeriod"), ViewVariables(VVAccess.ReadWrite)]
```
```suggestion
    [DataField("gracePeriod"), ViewVariables(VVAccess.ReadWrite)]
    [AutoNetworkedField]
```

DEATHB4DEFEAT commented on Content.Shared/SimpleStation14/Holograms/Components/HologramProjectedComponent.cs:
```diff
@@ -0,0 +1,92 @@
+    /// <remarks>
+    ///     Note that making this number larger than PVS is highly inadvisable, as the client will be stuck predicting the Hologram returning while the server confirms that they do not.
+    /// </remarks>
+    [DataField("projectorRange"), ViewVariables(VVAccess.ReadWrite)]
```
```suggestion
    [DataField("projectorRange"), ViewVariables(VVAccess.ReadWrite)]
    [AutoNetworkedField]
```

DEATHB4DEFEAT commented on Content.Shared/SimpleStation14/Holograms/Components/HologramProjectedComponent.cs:
```diff
@@ -0,0 +1,92 @@
+    /// <summary>
+    ///     The prototype of the effect to spawn for the Hologram's projection. Leave null to disable the visual projection effect.
+    /// </summary>
+    [DataField("effectPrototype", customTypeSerializer: typeof(PrototypeIdSerializer<EntityPrototype>))]
```
```suggestion
    [DataField("effectPrototype", customTypeSerializer: typeof(PrototypeIdSerializer<EntityPrototype>))]
    [AutoNetworkedField]
```

DEATHB4DEFEAT commented on Content.Shared/SimpleStation14/Holograms/Components/HologramProjectedComponent.cs:
```diff
@@ -0,0 +1,92 @@
+    ///     This provides a super cool effect of the Hologram only getting the visual information they technically should, but it's also a bit of a pain from a player perspective.
+    ///     Primarily used for the station AI.
+    /// </remarks>
+    [DataField("setEyeTarget"), ViewVariables(VVAccess.ReadWrite)]
```
```suggestion
    [DataField("setEyeTarget"), ViewVariables(VVAccess.ReadWrite)]
    [AutoNetworkedField]
```

DEATHB4DEFEAT commented on Content.Shared/SimpleStation14/Holograms/Components/HologramProjectedComponent.cs:
```diff
@@ -0,0 +1,92 @@
+    /// <summary>
+    ///     If set, the Hologram will only be able to be projected from this projector, simply ignoring all others.
+    /// </summary>
+    [ViewVariables(VVAccess.ReadOnly)]
```
```suggestion
    [ViewVariables(VVAccess.ReadOnly)]
    [AutoNetworkedField]
```

DEATHB4DEFEAT commented on Content.Shared/SimpleStation14/Holograms/Components/HologramProjectorActiveComponent.cs:
```diff
@@ -0,0 +1,10 @@
+///     Marks a hologram projector as active and working.
+/// </summary>
+[RegisterComponent, NetworkedComponent]
+public sealed class HologramProjectorActiveComponent : Component
+{ }
```
```suggestion
public sealed class HologramProjectorActiveComponent : Component
{

}
```

DEATHB4DEFEAT commented on Content.Shared/SimpleStation14/Holograms/Components/HologramServerLinkedComponent.cs:
```diff
@@ -0,0 +1,28 @@
+/// <remarks>
+///     This could be anything from a literal server, to an AICore, to the person a HoloParasite lives in.
+/// </remarks>
+[RegisterComponent]
```
```suggestion
[RegisterComponent, NetworkedComponent, AutoGenerateComponentState]
```

DEATHB4DEFEAT commented on Content.Shared/SimpleStation14/Holograms/Components/HologramServerLinkedComponent.cs:
```diff
@@ -0,0 +1,28 @@
+    ///     If true, it will be returned if it leaves the grid.
+    /// </summary>
+    [ViewVariables(VVAccess.ReadWrite)]
+    [DataField("gridBound")]
```
```suggestion
    [DataField("gridBound"), AutoNetworkedField]
```

DEATHB4DEFEAT commented on Content.Shared/SimpleStation14/Holograms/Components/HologramServerLinkedComponent.cs:
```diff
@@ -0,0 +1,28 @@
+    ///     This will be the lightbee if it's a lightbee hologram.
+    /// </remarks>
+    [ViewVariables(VVAccess.ReadOnly)]
+    public EntityUid? LinkedServer;
```
```suggestion
    [AutoNetworkedField]
    public EntityUid? LinkedServer;
```

DEATHB4DEFEAT commented on Content.Shared/SimpleStation14/Holograms/Systems/SharedHologramSystem.cs:
```diff
@@ -0,0 +1,120 @@
+
+        InitalizeServerLinked();
+    }
+
```
```suggestion


    public override void Initialize()
    {
        base.Initialize();

        SubscribeLocalEvent<HologramComponent, InteractionAttemptEvent>(OnHoloInteractionAttempt);
        SubscribeLocalEvent<HologramComponent, GettingInteractedWithAttemptEvent>(OnInteractionWithHoloAttempt);
        SubscribeLocalEvent<HologramComponent, StoreMobInItemContainerAttemptEvent>(OnStoreInContainerAttempt);
        SubscribeLocalEvent<HologramComponent, PreventCollideEvent>(OnHoloCollide);

        InitalizeServerLinked();
    }


```

DEATHB4DEFEAT commented on Content.Shared/SimpleStation14/Holograms/Systems/SharedHologramSystem.Projected.cs:
```diff
@@ -0,0 +1,263 @@
+{
+    [Dependency] private readonly IConfigurationManager _config = default!;
+
+    private void InitializeProjected()
```
Should this be called? Doesn't look like it is

DEATHB4DEFEAT commented on Content.Shared/SimpleStation14/Holograms/Systems/SharedHologramSystem.ServerLinked.cs:
```diff
@@ -0,0 +1,17 @@
+    {
+        SubscribeLocalEvent<HologramServerLinkedComponent, ChangedGridEvent>(OnGridChange);
+    }
+
```
```suggestion


```

DEATHB4DEFEAT commented on Content.Shared/SimpleStation14/Holograms/Systems/SharedHologramSystem.Projected.cs:
```diff
@@ -0,0 +1,263 @@
+{
+    [Dependency] private readonly IConfigurationManager _config = default!;
+
+    private void InitializeProjected()
+    {
+        SubscribeLocalEvent<HologramProjectedComponent, ComponentInit>(OnProjectedInit);
+    }
+
```
```suggestion

    private void InitializeProjected()
    {
        SubscribeLocalEvent<HologramProjectedComponent, ComponentInit>(OnProjectedInit);
    }


```

DEATHB4DEFEAT commented on Content.Shared/SimpleStation14/Holograms/Systems/SharedHologramSystem.Projected.cs:
```diff
@@ -0,0 +1,263 @@
+                TryKillHologram(hologram); // And if none is found, kill the hologram.
+                return;
+            }
+        }
```
```suggestion
        // If their last visited Projector is invalid ignoring occlusion and none is found
        if (!IsHoloProjectorValid(hologram, holoProjectedComp.CurProjector, 0, false) &&
            !TryGetHoloProjector(hologram, holoProjectedComp.ProjectorRange, out holoProjectedComp.CurProjector, holoProjectedComp, false)) 
        {
             // Kill the hologram.
            TryKillHologram(hologram);
            return;
        }
```

DEATHB4DEFEAT commented on Content.Shared/SimpleStation14/Holograms/Systems/SharedHologramSystem.Projected.cs:
```diff
@@ -0,0 +1,263 @@
+        }
+
+        // The two if statements above set the current projector, and kill if it's null, so we know it's not null moving forward.
+
```
```suggestion
```

DEATHB4DEFEAT commented on Content.Shared/SimpleStation14/Holograms/Systems/SharedHologramSystem.Projected.cs:
```diff
@@ -0,0 +1,263 @@
+    /// </summary>
+    /// <param name="coords">The coords to perform the check from.</param>
+    /// <param name="result">The UID of the projector, or null if no projectors are found.</param>
+    /// <param name="whiteList">A list of tags to check for on projectors, to determine if they're valid.</param>
```
```suggestion
    /// <param name="whiteList">An EntityWhitelist to check for on projectors to determine if they're valid.</param>
```

DEATHB4DEFEAT commented on Content.Shared/SimpleStation14/Holograms/Systems/SharedHologramSystem.Projected.cs:
```diff
@@ -0,0 +1,263 @@
+        if (!_timing.InPrediction) // TODOPark: HOLO Change this to run on the first prediction once it predicts reliably.
+        {
+            var holoPos = Transform(hologram).Coordinates;
+            _audio.Play(filename: "/Audio/SimpleStation14/Effects/Hologram/holo_off.ogg", playerFilter: Filter.Pvs(hologram), coordinates: holoPos, false);
```
Shouldn't be hard coded

DEATHB4DEFEAT commented on Content.Shared/SimpleStation14/Holograms/Systems/SharedHologramSystem.Projected.cs:
```diff
@@ -0,0 +1,263 @@
+            _popup.PopupCoordinates(Loc.GetString(PopupDisappearOther, ("name", meta.EntityName)), holoPos, Filter.PvsExcept(hologram), false, PopupType.MediumCaution);
+        }
+
+        // Does the do.
```
```suggestion
        // Teleports the hologram accurately
```

DEATHB4DEFEAT commented on Content.Shared/SimpleStation14/Holograms/Systems/SharedHologramSystem.Projected.cs:
```diff
@@ -0,0 +1,263 @@
+        // Plays the appearing effects.
+        if (!_timing.InPrediction)
+        {
+            _audio.PlayPvs("/Audio/SimpleStation14/Effects/Hologram/holo_on.ogg", hologram);
```
Shouldn't be hard coded

DEATHB4DEFEAT commented on Content.Client/SimpleStation14/Holograms/AcceptHologramEui.cs:
```diff
@@ -0,0 +1,44 @@
+    {
+        _window.Close();
+    }
+
```
```suggestion
```

DEATHB4DEFEAT commented on Content.Client/SimpleStation14/Holograms/AcceptHologramWindow.cs:
burn

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/Holograms/AcceptHologramEui.cs:
```diff
@@ -0,0 +1,32 @@
+{
+    private readonly HologramSystem _hologramSystem;
+    private readonly Mind.Mind _mind;
+
+    public AcceptHologramEui(Mind.Mind mind, HologramSystem hologramSys)
+    {
+        _mind = mind;
+        _hologramSystem = hologramSys;
+    }
+
```
```suggestion


    public AcceptHologramEui(Mind.Mind mind, HologramSystem hologramSys)
    {
        _mind = mind;
        _hologramSystem = hologramSys;
    }


```

DEATHB4DEFEAT commented on Resources/Locale/en-US/SimpleStation14/holograms.ftl:
```diff
@@ -0,0 +1,4 @@
+system-hologram-phasing-appear-self = You materialize into being!
+system-hologram-phasing-appear-others = {$name} materializes into being!
+system-hologram-phasing-disappear-others = {$name} dematerializes in a fizz!
```
Are holograms material?

DEATHB4DEFEAT commented on Resources/Prototypes/SimpleStation14/Entities/Body/Prototypes/scutter.yml:
```diff
@@ -0,0 +1,7 @@
+- type: body
+  id: Scutter
+  name: "scutter"
```
```suggestion
- type: body
  id: Scutter
  name: scutter
```

DEATHB4DEFEAT commented on Resources/Prototypes/SimpleStation14/Entities/Effects/hologram_effects.yml:
```diff
@@ -0,0 +1,23 @@
+  id: EffectHologramProjectionBeam
+  noSpawn: true
+  components:
+  # - type: PointLight
+  #   radius: 10.5
+  #   energy: 15
+  #   color: "#18abf5"
+  # - type: TimedDespawn
+  #   lifetime: 1
```
```suggestion
```

DEATHB4DEFEAT commented on Resources/Prototypes/SimpleStation14/Entities/Mobs/Player/hologram.yml:
```diff
@@ -0,0 +1,261 @@
+  #   damageContainer: Inorganic
+  - type: InteractionOutline
+  - type: Sprite
+    netsync: false
```
```suggestion
```

DEATHB4DEFEAT commented on Resources/Textures/SimpleStation14/Effects/hologram_projection_beam.rsi/meta.json:
```diff
@@ -0,0 +1,24 @@
+  "states": [
+    {
+      "name": "beam",
+      "delays": [
+        [
+          0.2,
+          0.2,
+          0.2,
+          0.2,
+          0.2,
+          0.2
+        ]
+      ]
```
```suggestion
      "delays": [ [ 0.2, 0.2, 0.2, 0.2, 0.2, 0.2 ] ]
```

DEATHB4DEFEAT commented on Resources/Textures/SimpleStation14/Effects/hologram_projection_beam.rsi/meta.json:
```diff
@@ -0,0 +1,24 @@
+{
+  "version": 1,
+  "license": "",
+  "copyright": "",
```
!

DEATHB4DEFEAT commented on Resources/Textures/SimpleStation14/Mobs/Silicon/scutter.rsi/meta.json:
```diff
@@ -0,0 +1,31 @@
+    },
+    {
+      "name": "shell",
+      "delays": [
+        [
+          1
+        ]
+      ]
```
```suggestion
```

Pspritechologist committed:
Work towards cleaning up, making the Eye better.

Pspritechologist committed:
Apply suggestions from code review

Co-authored-by: DEATHB4DEFEAT <77995199+DEATHB4DEFEAT@users.noreply.github.com>

Pspritechologist commented on Resources/Prototypes/SimpleStation14/Entities/Mobs/Player/hologram.yml:
```diff
@@ -0,0 +1,260 @@
+    isHardLight: false
+    collideTags:
+      - Wall
+      - Airlock
```
Make it so chairs are interactable but only when anchored.
Requesting a chair be screwed down so you can sit in it is very funny.

Pspritechologist commented:
A looping, very soft 'vwooom' sound effect that plays on whenever the hologram is currently *not* colliding with something they otherwise would be?

Pspritechologist commented on Content.Server/SimpleStation14/Holograms/Systems/HologramSystem.cs:
```diff
@@ -0,0 +1,189 @@
+
+        var pref = (HumanoidCharacterProfile) _prefs.GetPreferences(mind.UserId.Value).SelectedCharacter;
+
+        var mob = HoloFetchAndSpawn(pref, coords, "MobHologramProjected");
```
Very hardcoded :P

Pspritechologist committed:
Lots of using hardcoded values less, switched collision to a whitelist.

Should Holograms collide with windows?

Pspritechologist committed:
Making more things holograms

Pspritechologist committed:
CCTV System is in place for getting Holodisks... *

Getting some weird errors with the UI system that makes it not testable, but it works in theory.
It's super crummy regardless, but it would work for putting Holograms in game.

N0mad1c-K9 commented on Resources/Locale/en-US/SimpleStation14/holograms.ftl:
```diff
@@ -0,0 +1,4 @@
+system-hologram-phasing-appear-self = You materialize into being!
+system-hologram-phasing-appear-others = {$name} materializes into being!
+system-hologram-phasing-disappear-others = {$name} dematerializes in a fizz!
+system-hologram-phasing-death-self = You faze out of being!
```
```suggestion
system-hologram-phasing-death-self = You phaze out of being!
```

N0mad1c-K9 replied:
🤓
